[1단계 - 출석 구현] 띠용(주용은) 미션 제출합니다.

jye326 opened the pull request:
## 체크 리스트

- [ ] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [X] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
- [띠용 프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/3920)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [X] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

아쉬운 점이 꽤 있었다.

- 기본 입력 테스트 : 단위 테스트인지는 모르겠지만 미리 사용자의 입력값을 줄줄이 인자로 전달해 프로그램의 동작 과정 전체를 테스트하는 코드를 작성하고 싶었는데, 전달 방법을 몰라 실패했다.
- 모든 배열을 컬렉션으로 변경하지 않았다.
  -  이 부분은 정말 모든 배열을 ArrayList 등으로 변경해야 하는지 의문이다. 원소가 추가될 가능성이 없는 경우에도 List를 활용하는 것이 좋은 것일까?
- Enum 적용을 못했다.
  - 시간이 없었다. 어디에 어떻게 Enum 적용을 하면 좋을지 빠르게 떠오르지 않는다.
- 모든 원시값과 문자열을 포장한다.
  - 시간이 없었다. '모든' 이라고 하면 어디까지일까? 취향차인가?
- 일급 컬렉션을 쓴다.
  - 아직 개념도 잘 모르겠고, 필요한지도 모르겠다.
- 예외 처리를 똑바로 신경쓰지 못했다.
  -  시간이 없었다.
  - 일단 기능 구현하고 나중에 리팩토링 하자는 생각이었는데 생각해보니 다양한 상황에 대해 예외 테스트를 돌려보지도 않았고, 예외 상황에 대한 처리도 일관적이지 않다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

위의 아쉬운 점 위주로 봐주시면 좋겠습니다!

jye326 committed:
docs(README.md) : 1차 기능 요구사항 정리

jye326 committed:
test(domain) : 출석 상태 테스트 작성

jye326 committed:
feat(domain) : 출석 상태 작성

jye326 committed:
test(domain) : 출석 확인 테스트 작성

jye326 committed:
feat(domain) : 출석 확인 작성

- 출석, 지각, 결석 구현
- 중복 출석 방지 구현

jye326 committed:
test(domain) : 출석 수정 테스트 작성

jye326 committed:
feat(domain) : 출석 수정 작성

- 정상 수정 기능 구현
- 존재하지 않는 출석 수정 예외 처리

jye326 committed:
test(domain) : 출석 기록 출력 테스트 작성

- 결석 없는 경우
- 결석한 경우 추가

jye326 committed:
feat(domain) : 출석 기록 출력 기능 작성

- 결석 없는 경우
- 결석한 경우 추가

jye326 committed:
test(domain) : 출석 상태 출력 테스트 작성

- 출석, 지각, 결석별 횟수 출력

jye326 committed:
feat(domain) : 출석 상태 출력 기능 작성

- 출석, 지각, 결석별 횟수 출력

jye326 committed:
test(domain) : 제적 위험자 출력 테스트 작성

- 모든 제적 위험자 조회 결과 출력

jye326 committed:
feat(domain) : 제적 위험자 출력 기능 작성

- 모든 제적 위험자 조회 결과 출력

jye326 committed:
test(domain) : 제적 위험자 정렬 출력 테스트 작성

- 모든 제적 위험자 조회 결과 정렬 출력

jye326 committed:
feat(domain) : 제적 위험자 정렬 출력 기능 작성

- 모든 제적 위험자 조회 결과 정렬 출력

jye326 committed:
test(domain) : 출석 상태 현황 출력 테스트 작성

- 출성 상태 현황 출력

jye326 committed:
feat(domain) : 출석 상태 현황 출력 기능 작성

- 출성 상태 현황 출력

jye326 committed:
feat(controller) : 컨트롤러 기능 추가

- 출석 확인
- 출석 수정
- 종료

jye326 committed:
feat(app) : 메인 클래스 추가

jye326 committed:
refactor(domain) : 크루 조회 기능 분할

jye326 committed:
docs(attendance.csv) : 출석 기록 문서 추가

jye326 committed:
feat(controller) : 제적 위험자 확인 기능 작성

- 제적 위험자 출력
- 출력 형식 수정
  - "-" 구분자 출력

jye326 committed:
feat(controller) : 크루별 출력 기록 확인 기능 작성

jye326 committed:
fix(application) : 버그 수정

- 어제 날짜까지 결석 기록 업데이트
- 출석/지각/결석 출력 형식 통일
- 파일 입력 이후 결석 기록 업데이트

jye326 committed:
refactor(view) : 뷰 분리

- InputView
- OutputView

jye326 committed:
fix(domain) : 제적 위험자 필터링 버그 수정

- 위험 요소 없는 크루 출력 제외

jye326 committed:
refactor(view) : 뷰 분리

- FileInputView 분리

jye326 committed:
docs(README.md) : 리드미 수정

- 기능 요구사항 작성
- ToDo 작성

jye326 committed:
refactor(controller) : 리팩토링

- 메서드 분리
- 필드 추가
- 필드 생성 방식 변경

jye326 committed:
fix(test) : 테스트 코드 수정

- 날짜 지정 방식 변경 적용

jye326 committed:
refactor(domain) : 출력 형식 변경

- 출석 상태 출력 수정
 - 괄호 추가

jye326 committed:
refactor(view, domain) : 출력 형식 수정

jye326 committed:
refactor(app) : 리팩토링

- 함수 10라인 규칙 적용
- 네이밍 수정
- indent 규칙 적용
- DateGenerator 삭제

summerlunaa commented on README.md:
> 기본 입력 테스트 : 단위 테스트인지는 모르겠지만 미리 사용자의 입력값을 줄줄이 인자로 전달해 프로그램의 동작 과정 전체를 테스트하는 코드를 작성하고 싶었는데, 전달 방법을 몰라 실패했다.

아마 `인수 테스트`를 말씀하시는 것 같아요. 가볍게 찾아보시면 궁금증 해소에 도움이 될 것 같은데요, 추후 배우게 될 내용이니 너무 딥하게 파진 말아주세요! 이런 테스트도 있다~ 정도로만 이해하고 넘어가시면 좋을 것 같아요 :)

> 모든 원시값과 문자열을 포장한다. 시간이 없었다. '모든' 이라고 하면 어디까지일까? 취향차인가?
> 일급 컬렉션을 쓴다. 아직 개념도 잘 모르겠고, 필요한지도 모르겠다.

원시값과 문자열을 포장했을 때의 장점은 잘 알려져있습니다. 원시값을 사용했을 때보다 의미를 명확히 할 수 있고, 관련 로직을 해당 클래스 내에서 관리할 수 있고, 불변으로 만든다면 안정성도 보장되겠죠. 하지만 모든 설계에는 트레이드오프가 있습니다. 모든 원시값과 문자열을 포장했을 때 클래스가 과도하게 많아지면 관리해야 할 코드과 클래스도 그만큼 늘어나게 됩니다. 이런 부분들을 고려해서 판단하시면 될 것 같습니다~

시간이 많이 모자르셨죠 🥹 enum 클래스 활용과 예외 처리는 리뷰 반영 & 리팩터링 하면서 추가해보시죵~

summerlunaa commented on src/main/attendance.csv:
이 친구도 적절한 패키지 안에 넣어줄까요?

jye326 replied:
resource 패키지 생성하여 넣었습니다!

summerlunaa replied:
보통은 java 패키지와 같은 레벨에 resource 패키지를 두는 것 같아요~

jye326 replied:
이동하였습니다!

summerlunaa commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public void printCheckedAttendance(AllCrew allCrew, String name, String[] time){
+        System.out.println(allCrew.addCrewAttendanceByName(name, LocalDateTime.of(today.getYear(),
```
view에서는 add, modify 등의 비즈니스 로직을 호출하고 있고, 반대로 domain에서는 printXXX 등의 출력 로직을 갖고 있어요. 이렇게 뷰와 도메인이 강하게 결합되면 어떤 문제점이 발생할 수 있을까요? 두 계층을 어떻게 분리할 수 있을까요? 고민해보시고 리팩터링 부탁드립니다~

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,73 @@
+    private final UserInputView userInputView;
+    private final OutputView outputView;
+    private final AllCrew allCrew;
+    private final LocalDate date;
```
이 필드는 항상 오늘 날짜만 받고 있다면 의미를 명확하게 하기 위해 필드명을 todayDate라고 변경해도 좋겠네요.

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,73 @@
+        if (menuInput.equals("4")) {
+            checkDangerousCrew();
+        }
+        return !menuInput.matches("[Qq]");
```
정규식을 사용할 때 어떤 점에 주의해야 할까요? [블로그](https://blog2.deliwind.com/20200430/Java-%EC%A0%95%EA%B7%9C%ED%91%9C%ED%98%84%EC%8B%9D%EC%9D%84-%EC%82%AC%EC%9A%A9%ED%95%A0-%EB%95%8C-%EC%A3%BC%EC%9D%98%ED%95%A0-%EC%A0%90/)

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private boolean executeMenu(String menuInput) {
+        if (menuInput.equals("1")) {
```
Menu에 대한 상수들을 enum으로 관리해볼까요? :)

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,73 @@
+    private void modifyAttendance() {
+        String name = userInputView.askNickNameForModifyAttendanceInfo();
+        int day = userInputView.askDayForModifyAttendanceInfo();
+        String[] time = userInputView.askAttendanceTimeForModifyAttendance();
```
왜 배열보다 collection을 사용하라고 하는 걸까요?

summerlunaa commented on src/main/java/domain/AllCrew.java:
```diff
@@ -0,0 +1,71 @@
+
+    public boolean containsCrewName(String crewName) {
+        return allCrew.stream()
+                .anyMatch(crew -> crew.getName().equals(crewName));
```
이름을 가져오는 대신 객체에게 물어볼까요?

jye326 replied:
수정했습니다!

summerlunaa replied:
수정이 안 된 것 같아요 ㅎㅎ

jye326 replied:
반영이 안되어있었네요.. 죄송합니다!!

summerlunaa commented on src/main/java/domain/AllCrew.java:
```diff
@@ -0,0 +1,71 @@
+        return crew.update(dateTime);
+    }
+
+    public String printAttendanceHistory(String name, LocalDate lastDate) {
```
view에서 드린 리뷰와 동일합니다. domain에서 view 관련 로직들은 전부 제거해주세요~

summerlunaa commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,73 @@
+        if ((hour == startHour) && minute <= 30) {
+            return "지각";
+        }
+        return "결석";
```
상태 관련한 문자열, 지각 기준(5분, 30분) 등의 상수를 모아서 enum으로 관리하면 편하겠네요 👍

jye326 replied:
상태값, 출석 기준값 각각 상수화 하였습니다!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,158 @@
+                .filter(attendance ->
+                        attendance.isEqualDate(localDateTime))
+                .findFirst();
+        return sameDateAttendance.orElse(null);
```
빈 값에 대한 반환이 반드시 필요하다면 null 보다는 Optional을 사용하는 게 안전합니다. 가능하면 Optional도 필요한 곳에서만 사용하는 곳이 좋구요. 매번 빈 값인지 확인 후 사용해야 해서 값의 사용성을 떨어뜨리고 가독성을 해치거든요..!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,158 @@
+        return sameDateAttendance.orElse(null);
+    }
+
+    public String update(LocalDateTime newDateAndTime) {
```
업데이트 후 String이 결과값으로 나온다면 사용자는 이 값이 뭐라고 예측할 수 있을까요? 무슨 값인지 굉장히 헷갈릴 것 같아요. 다른 메서드들도 마찬가지구요. 메서드가 여러 역할을 하고 있는 건 아닌지 고민해보면 좋을 것 같습니다.

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    public void updateUntil(LocalDate lastDate) {
+        for (int i = 1; i <= lastDate.getDayOfMonth(); i++) {
```
`i`는 변수명으로 사용하지 말아주세요 🥹 더 구체적이고 명확한 이름을 붙여줍시다!

jye326 replied:
수정했습니다!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,158 @@
+        }
+        if (absentCount >= 2) {
+            return "경고";
+        }
```
여기도 enum으로 관리하면 편하겠네요~

jye326 replied:
반영했습니다!

summerlunaa commented on src/test/java/domain/CrewTest.java:
```diff
@@ -0,0 +1,187 @@
+                crew.update(LocalDateTime.of(2024, 12, 5, 10, 6)));
+    }
+
+    @DisplayName("출석 수정 예외")
```
어떤 예외 상황인지 더 구체적으로 작성해주시면 테스트를 더 파악하기 좋겠네요. 전체적으로 확인 부탁드립니다~
리팩토링 하실 때 커밋 메시지도 어떤 부분을 리팩토링 했는지 구체적으로 적어주시면 더 좋을 것 같습니다.

jye326 replied:
수정했습니다!

summerlunaa replied:
테스트가 누락된 곳이 많이 보이네요 😢 이번 미션의 키워드가 TDD인만큼 누락된 단위 테스트 모두 작성 부탁드리겠습니다.

summerlunaa reviewed (changes requested):
띠용 안녕하세요~ 파랑입니다 🐳 만나서 반갑습니다!
이번 미션은 요구사항도 많고 TDD도 적용하느라 쉽지 않았을 것 같은데 그래도 잘 구현해주셨습니다! 
MVC 패턴 적용과 객체의 역할&책임 분리에 대한 리뷰 중심으로 리팩토링 해보시면 좋을 것 같은데요, 이해가 안 되는 부분이 있거나 질문 있으시면 편하게 코멘트나 DM 남겨주세요. 
너무 고생하셨습니다 👍

jye326 committed:
refactor(resource) : attendance.csv 파일 이동

jye326 committed:
refactor(controller) : 필드 이름 수정

- date -> todayDate
 - 현재 날짜 의미가 더 잘 드러나도록 변경

jye326 committed:
refactor(app) : 도메인, 뷰 분리

- 출력 형식 포맷팅 등의 출력 관련 모든 로직 OutputView 로 이동
- 함수 반환형 변경 : 문자열 -> 필요 데이터 타입

jye326 committed:
refactor(controller) : 출력 형식 수정

- 시스템 출력 함수 호출 위치 컨트롤러로 변경

jye326 committed:
feat(constant) : 메뉴 옵션 값 상수화

- MenuOption enum 클래스 추가

jye326 committed:
fix(controller) : 시간 반환 타입 변경

배열 -> 컬렉션 변경

jye326 committed:
fix(domain) : 이름 필터링 조건 수정

- Crew의 이름을 getter로 가져와 비교 -> 비교 함수를 호출

jye326 committed:
feat(constant) : 상태값 상수화

- 출석 상태 관련 상수
- 출석 기준 관련 상수
- 경고 상태 및 기준 상수

jye326 committed:
refactor(domain, view) : 도메인, 뷰 분리

- 제적 위험자 확인 기능 domain, view 분리
- AllCrew
 - 크루 정렬과 제적 위험 크루 리스트 반환
- Crew
 - 현재 위험 상태 계산 및 반환
 - 불필요한 함수 제거
- OutputView
  - 포맷팅과 출력

jye326 committed:
refactor(domain) : 함수 이름 변경

- 미 출석 날짜 결석으로 채우는 함수 updateUntil -> updateAbsentUntil

jye326 committed:
test(domain) : 테스트 수정

- 도메인 뷰 분리에 따른 테스트 코드 수정
 - 뷰 의존성 제거
- 동일 기능 테스트 그룹화
- DisplayName 수정

jye326 committed:
fix(domain) : AllCrew 정렬 기능 수정

- 현재 날짜 인자 제거
 - 별도의 결석 정보 업데이트 함수
- 결석기준 내림차순, 이름 정렬 오름차순으로 수정

jye326 committed:
test(domain) : AllCrew 정렬 기능 테스트

- 결석+3회지각 기준으로 내림차순 정렬
- 이름 오름차순 정렬

jye326 reviewed (commented):
- **원시값과 문자열 포장**
    - **원시값 포장의 장점**
        1. 객체에게 자율성을 부여할 수 있습니다.
        2. 적절한 책임을 부여하여 유지보수성을 높일 수 있습니다.
            - 특히 원시 필드값에 대한 검증 로직이 필요한 경우, 해당 필드값을 포장하는 방법을 고려해볼 수 있습니다.
        3. 검증 로직은 생성자보다는 정적 팩토리 메서드에서 호출하는 편이 좋습니다.
            - 참고한 링크: https://devwriter.tistory.com/6
- **`attendance.csv` 파일 패키지 이동**
    - 해당 파일을 관련된 패키지로 이동하였습니다.
- **View와 Domain 간의 강한 결합 분리**
    - 기존에는 `Attendance` 내부에서 출력 형식 포맷팅 로직을 처리하고 있었습니다.
    - 이를 뷰 로직(`OutputView`)으로 이동하였습니다.
    - 테스트에 뷰 로직이 개입되지 않도록 주의.
        - 뷰 방식이 변경될 때마다 테스트 코드를 전부 수정해야 하는 문제가 발생했습니다.
        - 앞으로 테스트에 출력값을 포함시키면 안되겠다는 깨달음을 얻었습니다.
- **`AttendanceSystem.java`의 `LocalDate date` 의미**
    - 항상 오늘 날짜만 받는 용도라면, `todayDate`와 같이 좀 더 명확한 이름으로 변경을 검토하실 수 있습니다.
- **정규식 사용 시 주의점 수정**
    - `Pattern` 객체를 사용하도록 변경하였습니다.
- **Menu에 대한 상수를 enum으로 관리**
    - 관련 내용을 반영하였습니다.
- **배열보다 `Collection`을 사용해야 하는 이유**
    - 변경은 완료했지만, 예를 들어 `String[] times`처럼 단순히 시·분만 전달하는 용도로도 `Collection`을 사용하는 것이 좋은지에 대한 의문이 남습니다.
    - `Collection`은 다양한 기능을 제공하며, 크기 확장에도 용이하다는 장점이 있습니다. 배열과 Collection의 메모리 차이가 무시할 만 하며, 당장에 변경 계획이 없더라도, 추후 전달 인자가 늘어날 가능성을 고려하여 배열 대신 `Collection`을 사용하는 편이 낫다고 이해해도 괜찮을까요?
- **이름을 가져오는 대신 객체에게 직접 물어보기**
    - 해당 내용을 반영하였습니다.
- **Domain에서 View 관련 로직 전부 제거**
    - 관련 로직을 제거하였습니다.
- **지각, 5분·30분 등의 상수를 enum으로 관리**
    - 반영하였습니다.
- **빈값 반환이 필요한 경우 `null` 대신 `Optional` 사용**
    - 적용 완료하였습니다.
- **`Crew.update` 후 결과값이 `String`이면 혼동 가능 → 메서드 개선**
    - 수정하였습니다.
- **`i`와 같은 변수명 → 구체적인 이름으로 변경**
    - 의미 있는 이름(date)으로 변경하였습니다.
- **제적·면담 경고 → enum으로 관리**
    - 반영하였습니다.
- **`CrewTest` 출석 수정 예외의 `@DisplayName` 구체화**
    - 수정하였습니다.
- **커밋 메시지에서 리팩토링 상세히 기술**
    - 작성하였습니다.

summerlunaa commented on src/main/java/constant/AttendanceStatus.java:
```diff
@@ -0,0 +1,17 @@
+package constant;
+
+public enum AttendanceStatus {
+    ATTENDED("출석"),
+    LATE("지각"),
+    ABSENT("결석");
```
의미있는 상수들을 enum으로 분리하신 건 너무 잘하셨어요 👍 
다만 이 enum들을 단순히 상수로만 바라보지 않고, 역할과 책임을 가질 수 있는 하나의 도메인 객체로 보면 더 좋을 것 같아요.
AttendanceStatus라면 출석 상태와 관련된 일들을 할 수 있는거죠! 그럼 AllCrew, Crew, Attendance 등에 몰린 역할과 책임을 분리할 수 있을 것 같아요.(패키지도 domain으로 이동해야겠죠?) 필요하면 다른 클래스를 생성해도 괜찮구요.

jye326 replied:
지금까지 enum을 단순히 상수 묶음으로 인식하고 있었는데,  출석 상태 관련 로직을 담당하는 도메인 객체로 바라볼 수 있군요! 감사합니다!

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private boolean executeMenu(String menuInput) {
+        allCrew.updateAbsentHistory(todayDate.minusDays(1));
```
`updateAbsentHistory`는 왜 반복해서 호출하는 건가요?

jye326 replied:
프로그램이 진행되는 도중에 현재 날짜가 변경되어 중간에 결석 처리를 해야 하는 경우가 생길 수도 있겠다고 생각했으나, 현재 코드 상 프로그램 진행 중에는 날짜가 고정되어 있어 결석 업데이트 과정은 초기에 한 번으로 충분하다고 생각하여 수정했습니다.

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,93 @@
+
+    private boolean executeMenu(String menuInput) {
+        allCrew.updateAbsentHistory(todayDate.minusDays(1));
+        if (menuInput.equals(CHECK_ATTENDANCE.getValue())) {
```
menuInput을 string으로 비교하지 말고 MenuOption으로 만들어서 비교하면 더 쉽겠죵? 존재하지 않는 명령어가 들어왔는지 확인도 가능하구요.

jye326 replied:
String을 해당하는 값의 enum 객체로 반환하는 함수를 enum에 만드는 방법이 있었네요! 수정했습니다!

summerlunaa commented on src/main/java/controller/AttendanceSystem.java:
```diff
@@ -0,0 +1,93 @@
+    private void checkAttendance() {
+        String name = userInputView.askNickNameForCheckAttendance();
+        ArrayList<String> time = userInputView.askAttendanceTimeForCheckAttendance();
+        System.out.println(outputView.printCheckedAttendance(allCrew.addCrewAttendanceByName(name, LocalDateTime.of(todayDate.getYear(),
```
출력은 컨트롤러가 아닌 outputView에서 진행하시죠~

jye326 replied:
수정했습니다!

summerlunaa commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    public void printAttendanceHistory(AllCrew allCrew, String name) {
+        Crew crew = allCrew.getUpdatedCrew(name, today.minusDays(1));
```
출력 로직은 도메인에서 빠졌는데 뷰에서는 비즈니스 로직이 제거되지 않았네요 ㅠ
출석을 저장하고, 수정하고, 값을 불러오는 비즈니스 행위에 대한 호출은 모두 controller에서 관리하고 outputView에서는 결과만 받아 출력하는 걸로 수정해봅시다!!

jye326 replied:
제거하였습니다!

summerlunaa replied:
여전히 OutputView에서 비즈니스 로직을 수행하고 있네요 🥲 view에서 도메인의 값을 변경하거나 비즈니스 로직을 수행하게 되면 비즈니스 로직이 변경되었을 때 view까지 수정해야 합니다. 유지 보수도 어렵고, 로직을 확인하려면 controller부터 view, domain까지 전부 확인해야 하니 코드를 이해하기도 어렵습니다. 

```java    
// as-is
    public void printModifyAttendance(AllCrew allCrew, String name, LocalDateTime dateTime) {
        AttendanceUpdateResult attendanceUpdateResult = allCrew.modifyCrewAttendanceByName(name, dateTime);
        Attendance oldAttendance = attendanceUpdateResult.getOldAttendance();
        Attendance newAttendance = attendanceUpdateResult.getNewAttendance();
        System.out.println("\n" + getFormatedModifiedAttendance(oldAttendance, newAttendance));
    }
```

```java    
// to-be
    public void printModifyAttendance(Attendance oldAttendance, Attendance new Attendance) {
        System.out.println("\n" + getFormatedModifiedAttendance(oldAttendance, newAttendance));
    }
```

이런식으로 비즈니스 로직은 모두 제거해주세요. 

```java
    public void printAttendanceHistory(AllCrew allCrew, String name) {
        Crew crew = allCrew.findCrewByName(name);
        System.out.println(getAttendanceHistory(crew));
    }
```
조회하는 기능에서 위처럼 이름으로 Crew 찾아오는 것도 모두 비즈니스 로직입니다. 전체적으로 점검 후 수정해주세요.

summerlunaa commented on src/main/java/view/UserInputView.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Locale;
+import java.util.Scanner;
+
+// TODO : 입력값 검증 로직
```
todo 주석은 제때제때 해치우지 않으면 공허한 외침으로만 남습니다... PR 제출 전에 todo 주석은 웬만하면 지우는 게 나아요. 진짜 필요하다면 누가 언제까지 무슨 일을 해야하는지 구체적으로 작성해두는 게 좋습니다.

jye326 replied:
제거하였습니다..!

summerlunaa commented on src/main/java/view/UserInputView.java:
```diff
@@ -0,0 +1,67 @@
+
+    public ArrayList<String> askAttendanceTimeForModifyAttendance() {
+        System.out.println("언제로 변경하겠습니까?");
+        return new ArrayList<>(Arrays.asList(new Scanner(System.in).nextLine().split(":")));
```
> Collection은 다양한 기능을 제공하며, 크기 확장에도 용이하다는 장점이 있습니다. 배열과 Collection의 메모리 차이가 무시할 만 하며, 당장에 변경 계획이 없더라도, 추후 전달 인자가 늘어날 가능성을 고려하여 배열 대신 Collection을 사용하는 편이 낫다고 이해해도 괜찮을까요?

점점 기능이 확장되면서 inputView에서 반환받은 값들을 언제 어디서 사용하게 될지 모릅니다. 가능하면 다양한 기능을 사용할 수 있어 확장성이 좋은 컬렉션을 사용하는 게 낫겠죠. 또 배열보다는 컬렉션을 사용하는 것이 더 안전합니다. 이유는 한번 공부해보시면 좋을 것 같아요 ㅎㅎ

jye326 replied:
넵! 감사합니다!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,117 @@
+        attendanceInfo.remove(oldAttendance);
+        Attendance attendance = new Attendance(newDateAndTime);
+        attendanceInfo.add(attendance);
+        return List.of(oldAttendance, attendance);
```
과거의 값과 최신값을 리스트로 반환하면 이 메서드를 사용하는 입장에서 반환되는 값이 뭔지 알 수 있을까요? 수정해주세요~

jye326 replied:
사용자가 반환값을 파악할 수 있도록 수정 결과 클래스를 만들어 반환하도록 수정하였습니다!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    // 인자 : 지각 3회를 결석 1회로 간주한 결석 횟수
+    public String calculateWarningStatus() {
```
인자가 없어용. 그리고 반환하는 값을 Warning 클래스로 변경할까요? String으로 반환하면 view에서 말곤 쓸 수 없어서 범용성이 떨어져요. 패널티가 없는 상태가 필요하다면 Warning에 `NONE` 등의 값을 추가해볼 수 있겠네요.
리뷰 쓰다 보니 Warning이라는 클래스명도 모호하네요. 좀 더 명확하게 수정해보시죠~

jye326 replied:
주석 삭제 후 클래스 명 AbsentPenalty로 수정하였습니다! 감사합니다!

summerlunaa commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,68 @@
+
+public class Attendance {
+    private final LocalDateTime dateAndTime;
+    private final String state;
```
String으로 돌아다니는 상태값들을 전부 새로 만드신 enum으로 바꿔주세요~~~
`모든 원시값을 포장하라`는 요구사항이 왜 나왔는지 고민해보세요!

jye326 replied:
String 필드를 AttendanceStatus로 수정하였습니다!

summerlunaa commented on src/main/java/domain/AllCrew.java:
```diff
@@ -0,0 +1,76 @@
+        return crew.update(dateTime);
+    }
+
+    public Crew getUpdatedCrew(String name, LocalDate lastDate) {
```
조회하는 메서드인가요 수정하는 메서드인가요?

jye326 replied:
Crew를 불러오기 전에 해당 크루의 결석을 업데이트 해야 한다고 생각하여 만들어진 메서드입니다. 현재는 프로그램 동작 시 모든 크루의 결석을 업데이트 하므로 필요하지 않은 메서드라고 판단하여 삭제하였습니다!

summerlunaa commented on src/test/java/domain/CrewTest.java:
```diff
@@ -0,0 +1,307 @@
+            Attendance attendanceResult = crew.addAttendance(attendedTime);
+
+            //then
+            assertThat(attendanceResult).isNotNull();   // 출석 결과 반환 테스트
+            assertThat(attendanceResult.getDateAndTime()).isEqualTo(attendedTime);  // 저장된 출석 시간 테스트
+            assertThat(attendanceResult.getState()).isEqualTo(AttendanceStatus.ATTENDED.getStatus());   // 저장된 출석 상태 테스트
```
assertAll을 사용하는 것과 어떤 차이가 있을까요?

jye326 replied:
assertThat을 여러개 사용하는 것 과 달리 assertAll을 사용하면 이전 테스트가 실패하여도 각각의 테스트 결과를 확인할 수 있습니다! 수정하였습니다. 지난 주에 배웠던 부분인데 깜빡했습니다! 감사합니다!

summerlunaa replied:
적용 안 된 부분이 많네요~ 다른 곳들도 수정 부탁드립니다!

summerlunaa reviewed (changes requested):
띠용 안녕하세요~
리뷰 잘 반영해주셨습니다 👍 변경된 부분과 완벽하게 반영되지 않은 부분에 대해서 추가로 리뷰 남겨드렸어요. 다음 제출때는 누락된 테스트들도 채워주시면 더 좋겠습니다 :)
그리고 리뷰에 대한 코멘트는 [여기](https://github.com/woowacourse/java-attendance/pull/11#pullrequestreview-2635583063)처럼 몰아서 작성하지 마시고 코멘트마다 남겨주셔야 확인하고 대화 나누기가 편할 것 같아요~
조금만 더 힘내보시죠ㅎㅎㅎ 궁금한 점 있으면 편하게 DM 주셔도 괜찮습니다. 파이팅!

jye326 committed:
fix(domain) : 이미 존재하는 크루 이름 여부 확인 기능 수정

- 이름 수정 : containsCrewName -> isContainedCrewName
- getter를 통한 비교 -> 객체에서 호출

jye326 committed:
refactor(view) : 출력 로직 뷰 분리

- printCheckedAttendance 출력 기능 OutputView로 이동

jye326 committed:
refactor(controller) : AttendanceSystem 수정

- printCheckedAttendance 출력 기능 OutputView로 이동
- 메뉴 입력 값 비교 방식 변경
 - 문자열 비교 -> enum 객체 생성
 - 지원 하지 않는 문자 입력 시 에러 발생

jye326 committed:
delete(domain) : AllCrew 미사용 메서드 삭제

- getUpdatedCrew 삭제
- findCrewByName 접근 제어자 public 으로 변경

jye326 committed:
docs(attendance.csv) : 파일 경로 이동

jye326 committed:
fix(domain) : AttendanceStatus 수정

- 위지 이동 : constant -> domain
- 기능 추가 : 기존 Attendance 가 담당하던 출석 상태 확정 기능
- Attendance 필드 변경
 - String state -> AttendanceStatus

jye326 committed:
refactor(domain) : enum 클래스 경로 변경

constant -> domain

jye326 committed:
fix(view) : FileReader 경로 변경

jye326 committed:
fix(domain) : AllCrew update 반환형 수정

- List<Attendance> -> AttendanceUpdateResult

jye326 committed:
test(domain) : 기능 테스트 추가

- AllCrewTest
 - 이미 포함된 크루 이름 여부 테스트
- CrewTest
 - 크루 출석 정보 정렬 테스트
- MenuOptionTest
 - 객체-값 매칭 테스트

jye326 reviewed (commented):
수정했습니다! 감사합니다!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,116 @@
+        return name;
+    }
+
+    private boolean isAlreadyAttendedDay(LocalDateTime localDateTime) {
+        return attendanceInfo.stream()
+                .anyMatch(attendance ->
+                        attendance.isEqualDate(localDateTime));
+    }
```
테스트가 없네용

jye326 replied:
isAlreadyAttendedDay()는 CrewTest의 비정상 출석 시도 테스트 중 중복 출석 시도 상황에서 테스트를 진행하고 있습니다. private 메서드라 직접 호출하지 못해 해당 메서드를 사용하는 addAttendance()에서 해당하는 경우의 에러 메시지를 받는 방식으로 테스트 하였는데 다른 방법이 있을까요?

summerlunaa replied:
private 메서드였군요~ 확인했습니다!

summerlunaa commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,116 @@
+        return getStateCount(ABSENT);
+    }
+
+    private int getStateCount(AttendanceStatus status) {
+        return (int) attendanceInfo.stream()
+                .filter(a -> a.getStatus() == status)
+                .count();
+    }
+
+    private boolean containsDayOfMonth(int date) {
+        return attendanceInfo.stream()
+                .anyMatch(a -> a.getDayOfMonth() == date);
+    }
```
a... 👀

jye326 replied:
1. 의미가 드러나는 변수명 사용
2. getter보다 객체에게 물어보기
적용하여 수정하였습니다!

summerlunaa replied:
한 군데만 수정된 것 같아요. 

리뷰를 반영할 땐 코멘트가 달린 코드 뿐만 아니라 전체적으로 비슷하게 작성된 곳은 없는지 일괄적으로 확인해주시면 같은 리뷰를 중복으로 받지 않아도 되니 서로에게 좋습니다 :) 👍

summerlunaa commented on src/test/java/domain/AttendanceTest.java:
```diff
@@ -0,0 +1,104 @@
+        @Test
+        void test2() {
+            Attendance attendance = new Attendance(LocalDateTime.of(2024, 12, 3, 9, 55));
+            assertThat(attendance.getStatusValue()).isEqualTo("출석");
```
Enum에서 value가 다른 단어로 바뀌면 테스트 결과도 일일히 수정해줘야겠네요. 가능하면 Enum 값으로 비교해보세요.
그리고 단위 테스트라고 하려면 Attendance를 테스트하기 보다는 AttendanceStatusTest를 추가해서 `AttendanceStatus#checkAttendanceState`를 테스트하는 것이 더 좋을 것 같습니다. 둘 사이에 어떤 차이가 있을까요?

jye326 replied:
checkAttendanceState()가 기존에 Attendance에 있어서 AttendanceTest에 작성된 것을 수정하지 못했습니다. 말씀하신 대로 AttendanceStatusTest를 추가하고, 출력값이 아닌 enum값을 비교하였습니다.

테스트 코드가 훨씬 간결하고, 출력 형식에 상관없이 검증 할 수 있게 되었습니다.

summerlunaa commented on src/main/java/domain/AllCrew.java:
```diff
@@ -0,0 +1,70 @@
+            }
+        }
+        return allWarningCrew;
+    }
```
테스트가 없어요.

jye326 replied:
작성하였습니다!

summerlunaa commented on src/main/java/domain/AllCrew.java:
```diff
@@ -0,0 +1,70 @@
+        return crew.update(dateTime);
+    }
+
+    public Crew findCrewByName(String name) {
+        return allCrew.stream()
+                .filter(crew -> crew.isSameName(name))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Crew not found"));
+    }
```
여기도 테스트요~

jye326 replied:
추가하였습니다!

summerlunaa reviewed (changes requested):
띠용 안녕하세요~
이번에도 리뷰 잘 반영해주셨습니다 👍 누락된 단위 테스트 작성과 view/domain 분리 정도만 더 챙겨주시면 1단계는 마무리할 수 있을 것 같아요~
궁금한 점 있으면 편하게 DM 주셔도 괜찮습니다. 다음 리뷰 요청에는 머지할 예정이니 조금만 더 힘내주세요 💪

jye326 committed:
refactor(domain) : AllCrew 네이밍 수정

- 의미가 더 명확하도록 함수 이름 변경
 - getAllWarningCrew -> getAllAbsentPenaltyReceivedCrew
- 에러 메시지 수정

jye326 committed:
test(domain) : AllCrew 테스트 케이스 추가

- findCrewByName 테스트
 - 정상, 에러 상황
- getAllAbsentPenaltyReceivedCrew 테스트
 - 각 패널티 별 필터링 상황

jye326 committed:
fix(domain) : 특정 요일 출석 정보 존재 확인 기능 수정

- 변수명 변경
- getter -> 함수 호출 방식으로 변경

jye326 committed:
test(domain) : 출석, 크루 테스트 수정

- AttendanceTest
 - 출석 상태 테스트 코드 이동 -> AttendanceStatusTest
- CrewTest
 - 출력값 비교 -> enum 객체 비교

jye326 reviewed (commented):
enum을 왜 써야 하는지, 단위 테스트는 어떻게 작성하는지 등 정말 많이 배우게 된 것 같습니다!! 꼼꼼히 봐주셔서 정말 감사합니다!!!

jye326 committed:
refactor(view) : OutputView 비즈니스 로직 제거

- printAllDangerousCrew
 - 정렬 로직 컨트롤러로 이동
- printWarningInfo -> printAbsentPenaltyReceivedInfo (rename)
- printAttendanceHistory
 - 이름으로 크루 탐색 로직 컨트롤러로 이동
 - getAttendanceHistory(삭제됨) 와 병합
- 출석 정보 출력 포맷팅 로직 Attendance 도메인으로 이동
 - getFormattedAttended
 - getFormattedTimeAndState
- 크루별 출력 포맷팅 로직 Crew 도메인으로 이동
 - getFormatedAttendanceInfo
 - getFormatedAttendanceStateInfo
 - getFormatedWarningStatus
- 미사용 필드 LocalDate 제거

jye326 committed:
test(domain) : 테스트 수정

복수의 단일 테스트 assertAll 적용

summerlunaa commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,56 @@
+        return attendanceStatus.getStringValue();
+    }
+
+    public String getFormattedAttended() {
+        LocalDateTime dateAndTime = getDateAndTime();
+        return dateAndTime.format(DateTimeFormatter.ofPattern("MM월 dd일 "))
+                + dateAndTime.getDayOfWeek().getDisplayName(TextStyle.FULL, Locale.KOREAN) + " "
+                + getFormattedTimeAndState();
+    }
```
ㅠㅠ view에서 비즈니스 로직이 빠진 여파로 출력 로직이 다시 도메인으로 들어왔네용 🥲 일단 이 부분은 DM으로 좀 더 얘기 나눠보고 2단계에서 수정해보시죠!!

summerlunaa reviewed (approved):
띠용 안녕하세요~
리뷰 반영하느라 너무 고생하셨어요 👍 추가로 리팩토링 필요한 부분들은 2단계에서 반영해보시죠~ 다음 단계에서는 TDD 사이클을 따라가면서 단위 테스트 작성에 더 신경써보면 도움이 될 것 같습니다.
1단계는 여기서 마무리 하겠습니다. 수고하셨습니다!
